File: 11/container-with-most-water2.py
# LeetCode 11. Sparse Matrix Multiplication
# 就做「矩陣乘法」，兩矩陣裡，有很多0。這題若直接照著模擬，速度會慢。
class Solution:
    def multiply(self, mat1: List[List[int]], mat2: List[List[int]]) -> List[List[int]]:
        M, K, N = len(mat1), len(mat1[0]), len(mat2[0])
        ans = [[0] * N for _ in range(M)]
        for i in range(M):
            for k in range(K):  # 調一下迴圈的順序
                if mat1[i][k]==0: continue  # 避開有0的項，速度就變快了!
                for j in range(N):
                    ans[i][j] += mat1[i][k] * mat2[k][j]
        return ans

File: 11/container-with-most-water2.c
// LeetCode 311. Sparse Matrix Multiplication
// 進行 sparse 矩陣乘法，裡面有很多項是 0 哦
/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */
int** multiply(int** mat1, int mat1Size, int* mat1ColSize, int** mat2, int mat2Size, int* mat2ColSize, int* returnSize, int** returnColumnSizes){
    int M = mat1Size, K = mat2Size, N = mat2ColSize[0];
    int ** ans = (int**)malloc(sizeof(int*)*M);
    for(int i=0; i<M; i++){
        ans[i] = (int*) calloc(N, sizeof(int)); // 都會設為0
    }
    *returnColumnSizes = (int*) malloc(sizeof(int)*M);
    for(int i=0; i<M; i++){
        for(int k=0; k<K; k++){
            if(mat1[i][k]==0) continue; // 避開有0的項，因為0乘任何數都是0
            for(int j=0; j<N; j++){
                ans[i][j] += mat1[i][k] * mat2[k][j];
            }
        }
    }

    *returnSize = M;
    for(int i=0; i<M; i++){
        (*returnColumnSizes)[i] = N;
//printf("%d ", N);  
    }
    return ans;
}
